addOns/pscanrulesAlpha/src/test/java/org/zaproxy/zap/extension/pscanrulesAlpha/JSFunctionPassiveScannerUnitTest.java

```diff
@@ -0,0 +1,224 @@
+
+        // Given
+        String body = "";
+        HttpMessage msg = createHttpMessageWithRespBody(body, "text/html;charset=ISO-8859-1");
+
+        // When
+        rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+
+        // Then
+        assertThat(alertsRaised, empty());
+    }
+
+    @Test
+    public void shouldAlertGivenCustomPayloadFunctionMatch()
+            throws HttpMalformedHeaderException, URIException {
+        // Given
+        String body = "Some text <script>$badFunction()</script>\nLine 2\n";
+        HttpMessage msg = createHttpMessageWithRespBody(body, "text/html;charset=ISO-8859-1");
+        List<String> functions = Collections.singletonList("$badFunction");
+        JSFunctionPassiveScanner.setPayloadProvider(() -> functions);
+
+        // When
+        rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+
+        // Then
+        assertThat(alertsRaised, hasSize(1));
+        assertEquals(alertsRaised.get(0).getEvidence(), "$badFunction");
+    }
+
+    @Test
+    public void shouldNotAlertGivenMatchOutsideScript()
+            throws HttpMalformedHeaderException, URIException {
+        // Given
+        String body =
+                "<h1>Some text <script>Something innocent happening here</script></h1>\n"
+                        + "<b>You should not use bypassSecurityTrustHtml()</b>\n";
+        HttpMessage msg = createHttpMessageWithRespBody(body, "text/html;charset=ISO-8859-1");
+
+        // When
+        rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+
+        // Then
+        assertThat(alertsRaised, empty());
+    }
+
+    @Test
+    public void shouldAlertGivenMatchInSecondScript()
+            throws HttpMalformedHeaderException, URIException {
+        // Given
+        String body =
+                "<h1>Some text <script>Something innocent happening here</script></h1>\n"
+                        + "<p><b>Just some words going on</b>\n"
+                        + "<script>$eval()</script></p>\n";
+        HttpMessage msg = createHttpMessageWithRespBody(body, "text/html;charset=ISO-8859-1");
+
+        // When
+        rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+
+        // Then
+        assertThat(alertsRaised, hasSize(1));
+        assertEquals(alertsRaised.get(0).getEvidence(), "eval");
+    }
+
+    @Test
+    public void shouldAlertOnceGivenMultipleMatchesHTML()
+            throws HttpMalformedHeaderException, URIException {
+        // Given
+        String body =
+                "<h1>Some text <script>eval()</script></h1>\n"
+                        + "<p><b>Just some words going on</b>\n"
+                        + "<script>bypassSecurityTrustHtml()</script></p>\n";
+        HttpMessage msg = createHttpMessageWithRespBody(body, "text/html;charset=ISO-8859-1");
+
+        // When
+        rule.scanHttpResponseReceive(msg, -1, createSource(msg));
+
+        // Then
+        assertThat(alertsRaised, hasSize(1));
```

Comment:
Could assert the evidence here too (and following test).

Reply:
Tests revealed a problem with the html search logic. Added a check when cycling through HTML elements that if evidence is not empty to break.